### PR TITLE
added multi-softpatching support + OSD messages for patches (#9947)

### PR DESCRIPTION
**Pull request by eadmaster**
[followup](https://github.com/libretro/RetroArch/pull/11792)

Moved into a separate branch as suggested.

### CODE REVIEW

**Comment by jdgleaver** [comment]
@eadmaster Thanks, that's much better!

@twinaphex I think we can merge this now

**Comment by jdgleaver** [comment]
Oh, hold on - Travis has revealed one mistake. The usage of `basename()` is incorrect - let me mark it in the code...

File: tasks/task_patch.c [hunk 1]
```diff
@@ -628,6 +633,21 @@ static bool apply_patch_content(uint8_t **buf,
       free(ret_buf);
       *buf  = patched_content;
       *size = target_size;
+
+      /* Show an OSD message */
+      if (show_notification)
+      {
+         const char *patch_filename = basename(patch_path);
```
Comment (jdgleaver) [inline tasks/task_patch.c:640]: We cannot use `basename()` - it is not platform-independent.

We must use `path_basename()` instead.

Comment (eadmaster) [reply tasks/task_patch.c]: fixed.

Please when you have time take a look at this [other PR by me](https://github.com/libretro/snes9x/pull/255) .

**Comment by jdgleaver** [comment]
@eadmaster Great, that's fine now. But it still needs a rebase to squash that second commit:

```
git checkout master
git pull upstream master
git checkout multi_patch
git rebase master
git rebase -i HEAD~2
<enter fixup for the 'fixed Travis error' commit>
git push --force origin multi_patch
```

I will indeed look at your other PR as soon as I get the chance :)

**Comment by eadmaster** [comment]
Can't you git merge+squash in this case?

https://stackoverflow.com/questions/5308816/how-to-use-git-merge-squash

**Comment by jdgleaver** [comment]
@eadmaster Okay...

You broke the PR again with your last force push. I waited a week to see if you would fix it, but you didn't.

Please run the following commands:

```
git checkout multi_patch
git rebase master
git push --force origin multi_patch
```

That should make it mergeable again.

**Comment by eadmaster** [comment]
sorry, i'm in a bad relationship with `git` atm, hopefully it'll get better with practice. :-)

**Comment by hizzlekizzle** [comment]
woooo! balloons and streamers, guys :D

**Comment by i30817** [comment]
@eadmaster @jdgleaver @hizzlekizzle  this appears to show the notification of application of the patch  like this:

game.ips
game.ips3
game.ips2
game.ips1

Is this on purpose? Isn't it more user friendly going up to make sure the patches apply in the natural order? Or is it just a bug from using a message model of notifications?

**Comment by i30817** [comment]
Can a message push another message? That's how i worked around swing thread event bugs like that in the aeons past.

**Pull request merged**